test/src/vfs_test_helper.h

```diff
@@ -0,0 +1,370 @@
+   * filesystem's configuration
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param config Configuration parameters
+   * @param error Error parameter
+   * @return Status OK if successful
+   */
+  virtual Status prepare_config(
+      tiledb_config_t* config, tiledb_error_t* error) = 0;
+
+  /**
+   * Creates bucket / container if does not exist
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) = 0;
+
+  /**
+   * Removes bucket / container if exists
+   * Only for S3, Azure
+   * No-op otherwise
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) = 0;
+
+  /**
+   * Get the name of the filesystem's directory
+   *
+   * @return string directory name
+   */
+  virtual std::string temp_dir() = 0;
+
+  /* ********************************* */
+  /*             ATTRIBUTES            */
+  /* ********************************* */
+
+  /** The directory name of the associated filesystem. */
+  std::string vfs_helper_temp_dir_;
+
+  /** The vector of supported filesystems. */
+  std::vector<SupportedFs*> fs_vec;
+};
+
+/**
+ * This class provides support for
+ * the S3 filesystem.
+ */
+class SupportedFsS3 : public SupportedFs {
+ public:
+  /* ********************************* */
+  /*               API                 */
+  /* ********************************* */
+
+  /**
+   * Returns Status upon setting up the associated
+   * filesystem's configuration
+   *
+   * @param config Configuration parameters
+   * @return Status OK if successful
+   */
+  virtual Status prepare_config(tiledb_config_t* config, tiledb_error_t* error);
+
+  /**
+   * Creates bucket if does not exist
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * Removes bucket if exists
+   *
+   * @param ctx The TileDB context.
+   * @param vfs The VFS object.
+   * @return Status OK if successful
+   */
+  virtual Status close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs);
+
+  /**
+   * Get the name of the filesystem's directory
+   *
+   * @return string directory name
+   */
+  virtual std::string temp_dir();
+
+  /* ********************************* */
+  /*             ATTRIBUTES            */
+  /* ********************************* */
+
+  /** The directory prefix of the S3 filesystem. */
+  const std::string S3_PREFIX;
+
+  /** The bucket name for the S3 filesystem. */
+  const std::string S3_BUCKET;
+
+  /** The directory name of the S3 filesystem. */
+  const std::string S3_TEMP_DIR;
+
+  /** The directory name of the S3 filesystem. */
+  std::string vfs_helper_temp_dir;
+};
```

Comment:
- Initialize all of these within the default constructor.
- Add a trailing '_' to the names because they are member variables.
- I know this will be an annoying request, but these were previously named in all-caps because they were acting as C-style `#define` constants. While this can be still be constant if only initialized in the constructor's _member initializer list_, they should not be in all-caps. So let's rename `S3_PREFIX` to `s3_prefix_`, etc.